sip/sip-006-runtime-cost-assessment.md

```diff
@@ -0,0 +1,570 @@
+
+Status: Draft
+
+Type: Standard
+
+Created: 10/19/2019
+
+License: BSD 2-Clause
+
+# Abstract
+
+This document describes the measured costs and asymptotic costs
+assessed for the execution of Clarity code. This will not specify the
+_constants_ associated with those asymptotic cost functions. Those
+constants will necessarily be measured via benchmark harnesses and
+regression analyses. Furthermore, the _analysis_ cost associated with
+this code will not be covered by this proposal.
+
+# Measurements for Execution Cost
+
+Execution cost of a block of Clarity code is broken into 5 categories:
+
+1. Runtime cost: captures the number of cycles that a single
+   processor would require to process the Clarity block. This is a
+   _unitless_ metric, so it will not correspond directly to cycles,
+   but rather is meant to provide a basis for comparison between
+   different Clarity code blocks.
+2. Data write count: captures the number of independent writes
+   performed on the underlying data store (see SIP-004).
+3. Data read count: captures the number of independent reads
+   performed on the underlying data store.
+4. Data write length: the number of bytes written to the underlying
+   data store.
+5. Data read length: the number of bytes read from the underlying
+   data store.
+
+Importantly, these costs are used to set a _block limit_ for each
+block.  When it comes to selecting transactions for inclusion in a
+block, miners are free to make their own choices based on transaction
+fees, however, blocks may not exceed the _block limit_ (if they do so,
+the leader forfeits all rewards from the block).
+
+# Static versus Dynamic Cost Assessment
+
+Tracking the execution cost of a contract may be done either dynamically
+or statically. Dynamic cost assessment involves tracking, at the VM level,
+the various metrics as a contract is executed. Static cost assessment is
+performed via analysis of the contract source code, and is inherently
+a more pessimistic accounting of the execution cost: list operations
+are charged according to the _maximum_ size of the list (per the type
+annotations and inferences from the source code) and branching statements
+are charged according to the _maximum_ branch cost (per metric tracked, i.e.,
+if one branch performs 1 write and has a runtime cost of 1, and another
+branch performs 0 writes and has a runtime cost of 2, the whole statement
+will be assessed as having a maximum of 1 write and runtime cost of 2).
+
+# Costs of Common Operations
+
+### Variable Lookup
+
+Looking up variables in Clarity incurs a non-constant cost -- the stack
+depth _and_ the length of the variable name affect this cost.
+
+The stack depth affects the lookup cost because the variable must be
+checked for in each context on the stack. The variable name affects
+the cost because variables are stored in a HashMap, and determining
+the key associated with a variable name requires hashing that name (a
+linear operation).
```

Comment:
Do we want to consider the extra _O(log n)_ time complexity of querying the datastore's n-item B+-tree in this calculation?  As long as we're using sqlite as the underlying datastore, all loads/stores are at least _Omega(n log n)_ space and time.  However, the algorithms and data structures in SIP-004 do not require this -- an on-disk hash map will also work (which would not have this extra complexity).  I'm leaning towards designing the cost analysis to assume that the underlying datastore has the theoretically-best runtime (i.e. _O(1)_ load/store with an on-disk hashmap), but it will also mean that we'll need to drop sqlite at some point in the future.

Reply:
For now, no. The cost analysis does not include the runtime cost of either the MARF or the side-store used with the MARF -- it merely counts the number of READs/WRITEs to the MARF. That means that whatever integration of the measures we do will need to address that cost. On that front, I'd agree that we should be charging O(1).

This document and PR also needs to be updated. I think we should fix maximum name lengths in Clarity, precisely so that no compute cost depends on name lengths.